Serialize enums as strings

lukehutch opened the pull request:
Fixes #885 -- enums should not be serialized as ints (this is brittle, and could even be dangerous for protected-class identity data, as mentioned in that issue).

Unfortunately I tried to write unit tests, but the current enum tests are for the analyzer, so it's not possible to test this change to enums serialization in a test database, or client-server connection, without some additional test infrastructure being developed.

## Pre-launch Checklist

- [x] I read the [Contribute](https://docs.serverpod.dev/contribute) page and followed the process outlined there for submitting PRs.
- [x] This update contains only one single feature or bug fix and nothing else. (If you are submitting multiple fixes, please make multiple PRs.)
- [x] I read and followed the [Dart Style Guide](https://dart.dev/guides/language/effective-dart/style) and formatted the code with [dart format](https://dart.dev/tools/dart-format).
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`), and made sure that the documentation follows the same style as other Serverpod documentation. I checked spelling and grammar.
- [x] I added new tests to check the change I am making.
- [x] All existing and new tests are passing.
- [x] Any breaking changes are documented below.

If you need help, consider asking for advice on the [discussion board](https://github.com/serverpod/serverpod/discussions).

## Breaking changes

Also I didn't want to get into the details of updating the database migration code to migrate databases to this new model (the column type would have to change, and the ints would have to be rewritten as strings). Maybe this could just be labeled as a breaking change to prevent the need for writing this migration code.

vlidholt commented:
We will keep this update for v2 as it will break things for users.

Isakdl commented:
Thank you for the PR! This is a better approach than we have today, but I'm reluctant approving this right now as it is a breaking change and we would have to bump the major version.

Maybe we could let the user configure this instead? By default use int, can be configured to use the string name instead, @vlidholt what do you think?

lukehutch commented:
@vlidholt it won't break things for users any more than any other database migration requirement will. This just needs to have the appropriate database migration code written.

Also I assume enums are not widely used, given that they are not actually even documented yet.

What is your timeline to v2 though? I'm maintaining 5 PRs against upstream right now, and the number will only grow, which causes constant complexity for my development process...

tgrushka commented:
Yeah I would totally love to see this ASAP if can be user-configured. I'm building an enterprise dashboard and will be sure to lock my enums before release and not change the ordering, but having the names option (either at the enum yaml level or the server level) would be great for data safety. Also can make data recovery easier or if external tools are used to analyze the data in the database.

lukehutch commented:
Anybody is welcome to make this configurable -- I identified (I think) all the locations that need to be changed in the PR.

vlidholt commented:
We discussed this issue in the team meeting today. We will include it if it is configurable in the `config/generate.yaml` file. E.g. `serializeEnumValuesAsStrings: true`. For new projects this line would be included by default. Older projects would still require no change, and the update path would be more clear.

For version 2.0 we will set it to be true by default, and you would need to opt out by setting `serializeEnumValuesAsStrings: false`.

We will include this PR pretty quick if someone want to continue the work on this, otherwise we will try to get it in somewhere up to the 1.2 release. :)

lukehutch commented:
@vlidholt Currently `config/generator.yaml` is only read by `serverpod_cli`, not by the server library. The server library would need to read this file too, in order to know how to serialize to / deserialize from from the database. Do you really want the server to become runtime-dependent upon `config/generator.yaml`?

lukehutch commented:
@vlidholt The other alternative I'm exploring is a migration line like the following:

```sql
ALTER TABLE account_details ALTER COLUMN account_state TYPE TEXT USING CASE
    WHEN account_state=0 THEN 'active'
    WHEN account_state=1 THEN 'paused'
    WHEN account_state=3 then 'deleted'
END;
```

lukehutch commented:
@vlidholt I added the migration code to the PR. I can't figure out how to test this, because the `migrations` directory in my server project only contains migrations for the `serverpod` and `serverpod_auth` modules, not for my own protocol files.

lukehutch commented:
Ugh, I managed to accidentally check changes to generated files into the PR again -- see #1076. This is super annoying. I'll leave those in place so you can see the problem it is causing to have generated files in Serverpod main.

lukehutch commented on tools/serverpod_cli/lib/src/database/create_definition.dart:
```diff
@@ -21,6 +21,11 @@ DatabaseDefinition createDatabaseDefinitionFromEntities(
                   isNullable: column.name != 'id' && column.type.nullable,
+                  enumValueNames: !column.type.isEnum
+                      ? null
+                      : column.type.dartType?.element?.children
```
This part I'm not sure of -- I think this is the correct way to get enum value names from a `DartType`, but I have no idea if this is right, and I can't figure out how to get the migration output currently.

lukehutch commented:
@vlidholt I figured out I have to manually generate migrations using `serverpod migrate`. With that command, I get the error:

```console
Column <column_name> of table <table_name> was modified in a way that it must be deleted and recreated.
```

...so the migration code aborts before my migration logic kicks in.

The code is a good starting point though... would appreciate you taking a look. I don't want to keep polishing the migration code if you don't want to manually migrate enums.

vlidholt commented:
@lukehutch, `serverpod migrate` is still very much experimental. 😅 It will be part of the next release though! I don't think we will automatically migrate the enums, but instead support both integer and string options.

lukehutch commented:
@vlidholt Please take another look at the PR.

I added the static `GeneratorConfig.instance` (set by the constructor, which is ugly but I think necessary to minimize code changes). I assume `GeneratorConfig` will only be loaded once per call to `serverpod generate`, otherwise this would be a bad idea.

I needed `GeneratorConfig.instance` to avoid having to plumb the value of `GeneratorConfig` all over the `serverpod_cli` codebase -- I started doing that, and it was dozens of very ugly changes across many files to get the value where it was needed. With this static variable in place, the one place in the generator that needs the value of `GeneratorConfig.instance.serializeEnumValuesAsStrings` (`generator/types.dart`) has access to the value, without passing it everywhere.

I also added `Serverpod.instance.serializeEnumValuesAsStrings`, for the client to be able to access this value. This means the client now reads (and therefore depends upon) `generator.yaml` at runtime (which doesn't seem right to me, but it's what you requested).

lukehutch commented:
@vlidholt The latest PR seems to work flawlessly in my testing, both for client/server serialization and also database storage/retrieval (but maybe I missed something, please try it out).

lukehutch commented on packages/serverpod/lib/src/server/serverpod.dart:
```diff
@@ -106,6 +107,9 @@ class Serverpod {
   internal.RuntimeSettings get runtimeSettings => _runtimeSettings!;
 
+  /// Whether to serialize enums as Strings rather than as ints
+  late bool serializeEnumValuesAsStrings;
```
@vlidholt I figured out how to totally remove the runtime dependency upon `generator.yaml` -- see the latest commit.

The remaining issue surrounds `GeneratorConfig.instance`. This is set to the *most recent instantiation* of `GeneratorConfig`, which is ugly (but which in practice should not cause a problem, because there should not be two `GeneratorConfig`s being used at the same time, given that `serverpod_cli` only processes one generator task at a time).

lukehutch replied:
@vlidholt one more thing -- currently `lib/src/generated/log_entry.dart` fails to compile in the github tests. I can't run `tools/generate_all` on my project and then check in the results, because around 25 generated files differ significantly from the versions checked in to git (#1076). You really need to find a way to not include generated files in git.

vlidholt commented on packages/serverpod/lib/src/database/value_encoder.dart:
```diff
@@ -20,6 +20,22 @@ class ValueEncoder extends PostgresTextEncoder {
       return "'${input.uuid}'";
+    } else if (input is Enum) {
+      // Determine whether the Enum is serialized as an integer or as a String
+      var enumValueAsJson = (input as dynamic)?.toJson();
```
Is it not possible to match this on actual runtime types? `input as dynamic` feels fragile.

lukehutch replied:
I literally stole that snippet from Serverpod code 😂

It's in the code that serializes objects to JSON.

vlidholt commented on tools/serverpod_cli/lib/src/config/config.dart:
```diff
@@ -33,8 +33,15 @@ class GeneratorConfig {
+  static GeneratorConfig? _instance;
+
+  /// Get the current [GeneratorConfig] instance.
+  static GeneratorConfig? get instance => _instance;
```
`static GeneratorConfing get instance {
  assert(_instance != null, 'GeneratorConfig is not loaded.');
  ...
}

Would be better.

vlidholt commented on tools/serverpod_cli/lib/src/config/config.dart:
```diff
@@ -240,7 +262,12 @@ class GeneratorConfig {
-        extraClasses: extraClasses);
+        extraClasses: extraClasses,
+        serializeEnumValuesAsStrings: serializeEnumValuesAsStrings);
+
```
```dart
assert(_instance == null, 'GeneratorConfig should only be created once.');
```

lukehutch replied:
I added this to `GeneratorConfig.load` rather than the constructor, so that it's possible to still construct other `GeneratorConfig`s by calling the constructor directly, without setting `_instance`.

vlidholt reviewed (changes requested):
Pretty cool work, I think this could be a viable approach! :)

It would be nice to get some tests for this change, to make sure it works with both types of code generation options.

lukehutch commented:
Thanks, although I have already put way more time into this than I can justify right now (and I'm also traveling currently) -- any chance one of your guys can write the tests, and make the changes you suggested?

vlidholt commented:
@lukehutch, we will, but it may take a few weeks as it's vacation time in Sweden now. ☀️

lukehutch commented:
@vlidholt : @SandPod committed enum tests in #1244 -- are these tests sufficient to cover serializing enums as strings (this issue, #1080)?

lukehutch commented:
For the record, this is the code that is generated for an enum, with this patch:

The enum:

```dart
enum: FileType
values:
  - text
  - photo
  - video
  - audio
```

The generated code:

```dart
/* AUTOMATICALLY GENERATED CODE DO NOT MODIFY */
/*   To generate run: "serverpod generate"    */

// ignore_for_file: library_private_types_in_public_api
// ignore_for_file: public_member_api_docs
// ignore_for_file: implementation_imports

// ignore_for_file: no_leading_underscores_for_library_prefixes
import 'package:serverpod_client/serverpod_client.dart' as _i1;

enum FileType with _i1.SerializableEntity {
  text,
  photo,
  video,
  audio;

  static final _nameToValue = <String, FileType>{
    for (var value in values) value.name: value
  };

  static FileType? fromJson(String name) {
    return _nameToValue[name];
  }

  @override
  String toJson() => name;
}
```

lukehutch commented:
@SandPod @Isakdl any chance this PR could make it into the work you're doing with expressions? I really want to reduce the delta between my forked Serverpod and the upstream version. This has working tests.

lukehutch commented:
Closing in favor of #1518
